File: spatial-index/src/main/java/lva/spatialindex/index/Distributions.java
package lva.spatialindex.index;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

import static java.util.stream.Collectors.toList;
import static lva.spatialindex.index.Entry.margin;

/**
 * @author vlitvinenko
 */
class Distributions {
    private Distributions() {}

    static class GroupPair {
        List<Entry> group1 = new ArrayList<>();
        List<Entry> group2 = new ArrayList<>();
    }

    static int getGroupMargins(List<List<GroupPair>> distributionGroups) {
        return distributionGroups.stream().mapToInt(Distributions::margins).sum();
    }

    static List<List<GroupPair>> getDistributionGroups(List<Entry> entries,
                                                       Collection<Comparator<Entry>> comparators) {

        Function<Comparator<Entry>, List<Entry>> sorted = cmp ->
                entries.stream().sorted(cmp).collect(toList());

        return comparators.stream().map(sorted)
                .map(Distributions::distributions)
                .collect(toList());
    }

    private static int margins(List<GroupPair> groups) {
        return groups.stream().mapToInt(g -> margin(g.group1) + margin(g.group2)).sum();
    }

    private static List<GroupPair> distributions(List<Entry> entries) {
        List<GroupPair> groups = new ArrayList<>();
        for (int k = 0; k < (Node.MAX_ENTRIES - 2 * Node.MIN_ENTRIES + 2); k++) {
            GroupPair pair = new GroupPair();
            pair.group1 = entries.subList(0, Node.MIN_ENTRIES + k);
            pair.group2 = entries.subList(Node.MIN_ENTRIES + k, entries.size());
            groups.add(pair);
        }
        return groups;
    }

}

File: spatial-index/src/main/java/lva/spatialindex/index/Distributions.kt
package lva.spatialindex.index

import lva.spatialindex.index.Entry.Companion.margin

/**
 * @author vlitvinenko
 */

internal typealias GroupPair = Pair<List<Entry>, List<Entry>>
internal typealias DistributionGroups = List<GroupPair>

internal fun getGroupMargins(distributionGroups: List<DistributionGroups>): Int =
    distributionGroups.sumOf { margins(it) }

internal fun getDistributionGroups(entries: List<Entry>, comparators: Collection<Comparator<Entry>>): List<DistributionGroups> =
    comparators.asSequence()
        .map { entries.sortedWith(it) }
        .map { distributions(it) }
        .toList()

private fun margins(groups: DistributionGroups): Int =
    groups.sumOf { margin(it.first) + margin(it.second) }

private fun distributions(entries: List<Entry>): DistributionGroups {
    val groups = arrayListOf<GroupPair>()
    for (k in 0 until Node.MAX_ENTRIES - 2 * Node.MIN_ENTRIES + 2) {
        val pair = GroupPair(
            entries.subList(0, Node.MIN_ENTRIES + k),
            entries.subList(Node.MIN_ENTRIES + k, entries.size)
        )
        groups.add(pair)
    }
    return groups
}
